Implement `managedChecksums` in ivy

jvican opened the pull request:
Managed checksums tells ivy to forget about checking checksums for jar
files and just persist them in the cache.

The user that enables that option will take care of verifying they are
correct.

This is a big deal because:
  
1. Ivy takes *a lot of time* checking timestamps for big jars, and does it
sequentially. The user (sbt) can do better by running these checks in parallel,
speeding up the whole resolution process!
2. The fact that the sha files are not present in the cache means that build
tools cannot check if a checksum is correct without preprocessing the jar.

Note that the user is responsible for keeping the cache consistent. If the
checksum is not correct, the user should report it and overwrite the
checksum file with the valid digest.

jvican committed:
Add custom checksum checks to ivy

jvican committed:
Fix bug in updateOptionsFormat

jvican committed:
Add checksum to ModuleID

A checksum in a `ModuleID` ensures that the checksum of the resolved
module id matches. This checksum check is done at the same time ivy
checks it.

jvican committed:
Add `managedChecksums` to `UpdateOptions`

jvican committed:
Implement `managedChecksums` in ivy

Managed checksums tells ivy to forget about checking checksums for jar
files and just persist them in the cache.

The user that enables that option will take care of verifying they are
correct.

This is a big deal because:

1. Ivy takes *a lot of time* checking timestamps for big jars, and does
   it sequentially. The user (sbt) can do better by running these checks
   in parallel, speeding up the whole resolution process!
2. The fact that the sha files are not present in the cache means that
   build tools cannot check if a checksum is correct without preprocessing
   the jar.

Note that the user is responsible for keeping the cache consistent. If
the checksum is not correct, the user should report it and overwrite the
checksum file with the valid digest.

jvican commented:
Review by @eed3si9n.

jvican commented:
I forgot to mention that this also adds a field `checksum` to `ModuleID`, which is super useful for the dependency lock file and for those users that want to explicitly say what the checksum a module is (a la Pants).

jvican committed:
Make explicit where we deviate from ivy behaviour

jvican commented:
Tests are green.

jvican committed:
Move `managedChecksums` to InlineIvyConfiguration

`UpdateOptions` is only meant for user-provided options. Since this
option should not be exposed to users, it's moved to inline ivy
configuration, whose parameters require information about the build and
is usually defined project-wide instead of `Global`-wide.

eed3si9n commented on librarymanagement/src/main/contraband/librarymanagement.json:
```diff
@@ -267,23 +267,34 @@
+        { "name": "extraAttributes",   "type": "Map[String, String]",                    "default": "Map.empty",                        "since": "0.0.1" },
+        { "name": "crossVersion",      "type": "sbt.librarymanagement.CrossVersion",     "default": "sbt.librarymanagement.Disabled()", "since": "0.0.1" },
+        { "name": "branchName",        "type": "Option[String]",                         "default": "None",                             "since": "0.0.1" },
+        { "name": "checksum",          "type": "Option[sbt.librarymanagement.Checksum]", "default": "None",                             "since": "0.0.1" }
```
I think this is at a wrong layer. First, a module may consist of many artifacts.
Second, it's an addressing mechanism, so you won't know the checksum up front.

eed3si9n reviewed (changes requested):
As noted in the Gitter chat, a better optimization around checksum is to utilize HTTP headers when available (Maven Central provides `X-Checksum-SHA1`).

Also ModuleID is not the right place, I think.

jvican committed:
Move checksum to artifact

This is a more elegant solution than before since module IDs can bring
different artifacts (in updateClassifiers, for instance).

eed3si9n commented:
Merged! Thanks for the contribution.

jvican commented:
I have just opened an issue to keep track of your suggestion: https://github.com/sbt/librarymanagement/issues/108.
